rfcs/shared_api_v1/src/main.rs

```diff
@@ -0,0 +1,507 @@
+    ///
+    /// This is where all the action happens.  This async method can make asynchronous calls
+    /// against either (1) the global state (remote object), or (2) the injector.
+    /// In either case, it is registering a continuation to respond to the completion of an RPC
+    /// in the host (coordinator) or guest respectively.
+    fn handle_event<I: Instrumentor>(g: &mut Remoteable<Self::Glob>, p: &mut Self::Proc, t: 
+                                    &mut Self::Thrd, i : &mut I, e : Event) -> ();
+        //  TODO: in the future each event handled may return a result to the instrumentor 
+        // which changes its configuration: for example, subscribing or unsubscribing to 
+        // categories of events.
+
+    // Optionally override this as an optimization.  By default it will send the event 
+    // to the respective
+    // fn handle_global_event<I: Injector>(e : Event, i: I, g: Self::Glob) -> ();
+}
+
+//--------------------------------------------------------------
+// TODO / unfinished
+//--------------------------------------------------------------
+
+/// Thread ID
+pub type TID = Pid;
+
+// TODO: replace this with whatever is most idiomatic.
+pub type SerializedVal = String;
+
+// The value returned in RAX on x86_64:
+pub type SysCallRet = i64;
+
+/// The 6 arguments of a syscall, raw untyped version.
+///
+/// TODO: Use a helper function to convert to a structured Syscall+Args enum.
+#[derive(PartialEq, Debug, Eq, Hash, Clone)]
+pub struct SysArgs {
+    arg0: u64,
+    arg1: u64,
+    arg2: u64,
+    arg3: u64,
+    arg4: u64,
+    arg5: u64,
+}
+
+#[derive(PartialEq, Debug, Eq, Hash, Clone)]
+pub enum Instr {
+    RDTSC,    
+    CPUID,
+    // Future, no current way to trap:
+    // RDRAND,
+    // HTX instructions...
+}
+
+// TODO Grab this enum from the lib:
+pub type SysNo = u64;
+pub type SigNo = u64;
+pub type FunAddr = u64;
+
+//--------------------------------------------------------------
+// Example tool: a simple counter.
+
+/// A counter tool, keeping the simplest possible global state: a number.
+#[derive(PartialEq, Debug, Eq, Hash, Clone)]
+pub struct CounterTool {}
+// AUDIT: is this singleton type idiomatic?
+
+/// Type-safe wrapper-methods for RPC calls.  These could be generated.
+impl CounterTool {    
+    fn incr<I : Instrumentor>(g: &mut Remoteable<u64>, i : &mut I, x : u64) -> () {
+       Self::exec_rpc(g, IncrMsg(x), i);
+    }
+}
+
+#[derive(PartialEq, Debug, Eq, Hash, Clone, Deserialize)] 
+pub struct IncrMsg(u64);
+
+// [derive(Serialize,Deserialize)]
+// [serde(tag = "type")]
+
+impl Serialize for IncrMsg {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S : serde::Serializer,
+    {        
+        match self {
+            IncrMsg(n) => n.serialize(serializer)
+        }
+    }        
+}
+
+impl SystraceTool for CounterTool {
+    type Glob = u64;    
+    type GlobMethodArgs = IncrMsg;
+    type GlobMethodResult = ();    
```

Comment:
@rrnewton How can one inspect the global state 1) from a tool and 2) from the guest? If it's via `receive_rpc` to retrieve certain fields of a struct, or `u64` in this case, specifying `GlobMethodResult` might be tricky. Or can we have a `fetch_global_state () -> &Glob` ?

Reply:
Well, only the Glob RPC methods should be able to directly access the `Glob` type.  The `handle_event` callbacks should be forced to send RPCs, irrespective of whether they happen to execute in the guest address space or in the centralized tracer/daemon.

Reply:
Alright. If we decide to encode this as separate RPC's to retrieve specific fields of a struct, or `u64` in this case, specifying `GlobMethodResult` might be tricky? Or can we have a `fetch_Glob () -> &Glob`, which would be a RPC call, but would just return a read-only `Glob`.